In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
promo=pd.read_excel('Promotion.xlsx')
promo.head()

Promotion Id       Date  Discount  ProductId
0   NzjwLJ2016 2019-10-01      0.05      10566
1   pK8KLJ2016 2019-10-02      0.08       1924
2   PmHuil2016 2019-10-03      0.10        798
3   IdZiv82016 2019-10-04      0.05        956
4   EN4jLJ2016 2019-10-05      0.10      10616

In [3]:
promo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Promotion Id  61 non-null     object        
 1   Date          61 non-null     datetime64[ns]
 2   Discount      61 non-null     float64       
 3   ProductId     61 non-null     int64         
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 2.0+ KB


In [4]:
sales=pd.read_csv('Sales_data.csv')
sales.event_date=pd.to_datetime(sales.event_date,dayfirst=True)

In [5]:
sales.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 160000 entries, 0 to 159999
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   user_id         160000 non-null  int64         
 1   xyz             160000 non-null  object        
 2   event_date      160000 non-null  datetime64[ns]
 3   Day_of_Week     160000 non-null  object        
 4   Channel         160000 non-null  object        
 5   event_time      160000 non-null  object        
 6   event_hour      160000 non-null  int64         
 7   event_timezone  160000 non-null  object        
 8   event_type      160000 non-null  object        
 9   product_id      160000 non-null  int64         
 10  category_id     160000 non-null  int64         
 11  category        160000 non-null  object        
 12  sub_category1   160000 non-null  object        
 13  sub_category2   160000 non-null  object        
 14  brand           160000 non-null  obj

# Special Promotion Analysis

In [6]:
# Merging the tables to get all the events linked to promoted products on the respective promoted date.
promoted_events=sales.merge(promo, how='right',left_on=['event_date','product_id'],right_on=['Date','ProductId'])
promoted_events

user_id         xyz event_date Day_of_Week  Channel event_time  \
0    1403209.0  10-01-2019 2019-10-01     Tuesday  Browser   08:26:17   
1    1589232.0  10-01-2019 2019-10-01     Tuesday  Browser   17:35:46   
2    1599426.0  10-01-2019 2019-10-01     Tuesday      App   15:07:14   
3     309543.0  10-01-2019 2019-10-01     Tuesday      App   17:49:00   
4     304558.0  10-01-2019 2019-10-01     Tuesday  Browser   05:35:35   
..         ...         ...        ...         ...      ...        ...   
314  2072236.0  11/28/2019 2019-11-28    Thursday      App   07:40:08   
315  1946271.0  11/28/2019 2019-11-28    Thursday  Browser   09:25:38   
316  1569465.0  11/28/2019 2019-11-28    Thursday  Browser   09:39:29   
317        NaN         NaN        NaT         NaN      NaN        NaN   
318  3020492.0  11/30/2019 2019-11-30    Saturday      App   17:31:36   

     event_hour event_timezone event_type  product_id  ...  sub_category2  \
0           8.0            UTC       View     10566.0  ...      headphone   
1          17.0            UTC   Purchase     10566.0  ...      headphone   
2          15.0            UTC       View     10566.0  ...      headphone   
3          17.0            UTC       View     10566.0  ...      headphone   
4           5.0            UTC   Purchase     10566.0  ...      headphone   
..          ...            ...        ...         ...  ...            ...   
314         7.0            UTC       Cart       871.0  ...     smartphone   
315         9.0            UTC       View       871.0  ...     smartphone   
316         9.0            UTC       View       871.0  ...     smartphone   
317         NaN            NaN        NaN         NaN  ...            NaN   
318        17.0            UTC       View       846.0  ...     smartphone   

       brand    price user_session State  User_Score  Promotion Id       Date  \
0      Apple   161.93    1015299.0    UT         1.0    NzjwLJ2016 2019-10-01   
1      Apple   161.93    8877683.0    IA         4.0    NzjwLJ2016 2019-10-01   
2      Apple   161.93   13550157.0    MN         4.0    NzjwLJ2016 2019-10-01   
3      Apple   161.93    3839051.0    NH         3.0    NzjwLJ2016 2019-10-01   
4      Apple   161.93    8976936.0    MT         2.0    NzjwLJ2016 2019-10-01   
..       ...      ...          ...   ...         ...           ...        ...   
314    Apple   916.37    2741471.0    MA         3.0    c1JgDc2016 2019-11-28   
315    Apple   916.37    6551877.0    KY         1.0    c1JgDc2016 2019-11-28   
316    Apple   916.37   12803672.0    VA         2.0    c1JgDc2016 2019-11-28   
317      NaN      NaN          NaN   NaN         NaN    eJML432016 2019-11-29   
318  Samsung  1158.03     627681.0    ID         2.0    eJML432016 2019-11-30   

     Discount ProductId  
0        0.05     10566  
1        0.05     10566  
2        0.05     10566  
3        0.05     10566  
4        0.05     10566  
..        ...       ...  
314      0.15       871  
315      0.15       871  
316      0.15       871  
317      0.05      7431  
318      0.30       846  

[319 rows x 23 columns]

### EDA on Special Promotions 

In [7]:
# Total Promotions
promoted_events['Promotion Id'].nunique()

16

In [8]:
# Total products promoted
promo.ProductId.nunique()

61

 With 61 promoted products for two months we got only 319 intearctions consisting of view, purchase and cart.

In [45]:
# Products that failed to gather any impact with the special promotion
promoted_events[promoted_events.user_session.isna()]['ProductId'].values

''' 25 out of 61 promoted products couldnt get any interaction (no view, no cart and no purchase event) even with special promotions.''' 

array([ 1924,   798,   956, 10616,  1391, 25413, 15351,   893, 11323,
       14213, 14002, 17075,  7457, 11304,  8292,   984, 14068,  6928,
        5536,  6794,  3347, 24259, 23742, 26016,  7431], dtype=int64)

In [46]:
# Category of above found Products 
promoted_events[promoted_events.user_session.isna()][['ProductId']].merge(sales,how='left',left_on='ProductId',right_on='product_id').category.unique() 

array(['Computers', 'Electronics', 'Construction', 'Appliances',
       'Apparel'], dtype=object)

In [10]:
# Brand wise total special promotion applied and total events/interactions happened.
def fun(x):
    return x.unique()
brand_prodcat_wise_interactions=promoted_events.groupby('brand').agg(Product_category=('category',fun),Total_Promotions_applied=('Promotion Id','nunique'),Total_interactions=('event_type','size'))\
.sort_values(by='Total_Promotions_applied',ascending=False).reset_index()

brand_prodcat_wise_interactions

brand           Product_category  Total_Promotions_applied  \
0     Apple              [Electronics]                         6   
1   Samsung  [Electronics, Appliances]                         6   
2        Lg  [Electronics, Appliances]                         4   
3      Acer                [Computers]                         2   
4    Huawei              [Electronics]                         2   
5      Oppo              [Electronics]                         2   
6    Xiaomi              [Electronics]                         2   
7     Artel              [Electronics]                         1   
8     Haier               [Appliances]                         1   
9   Indesit               [Appliances]                         1   
10   Lenovo                [Computers]                         1   
11    Meizu              [Electronics]                         1   
12  Philips               [Appliances]                         1   
13     Sony              [Electronics]                         1   

    Total_interactions  
0                  122  
1                  103  
2                    8  
3                    4  
4                   14  
5                   17  
6                   16  
7                    2  
8                    1  
9                    2  
10                   2  
11                   1  
12                   1  
13                   1

- So only electronics, appliances and computers category's products had impact of special promotion unlike some products (25) belonging to appliances,construction and apparel that did not gather any impact even with spceial promotions.

In [12]:
# Product- brand- category- channel- event type wise total interactions/events observed

Event_and_TotalPromo_summary=promoted_events.groupby(['product_id','brand','category','Channel','event_type']).agg(Total_events=('user_session','size')).reset_index()
Event_and_TotalPromo_summary

product_id    brand     category  Channel event_type  Total_events
0         11.0  Samsung  Electronics      App       View             2
1        300.0       Lg  Electronics      App       View             1
2        300.0       Lg  Electronics  Browser       View             1
3        448.0    Apple  Electronics      App       Cart             2
4        448.0    Apple  Electronics      App   Purchase             1
..         ...      ...          ...      ...        ...           ...
93     11017.0    Apple  Electronics      App   Purchase             2
94     11017.0    Apple  Electronics      App       View             3
95     11017.0    Apple  Electronics  Browser       Cart             1
96     11017.0    Apple  Electronics  Browser       View             4
97     11618.0  Philips   Appliances      App       View             1

[98 rows x 6 columns]

### Impact Analysis

The impact of promotion is observed using the Conversion Rate (CR)

Calculating a Baseline CR 

-  To make sure the impact observed is due to promotion only not because of some other factors.

In [49]:
baseline_cr=sales[sales.product_id.isin(promo.ProductId)]
baseline_cr.head()


user_id         xyz event_date Day_of_Week  Channel event_time  event_hour  \
3    38878  11/17/2019 2019-11-17      Sunday  Browser   06:20:37           6   
4    47492  11/16/2019 2019-11-16    Saturday  Browser   03:07:19           3   
5    48899  10-10-2019 2019-10-10    Thursday  Browser   05:56:38           5   
6    49983  10/27/2019 2019-10-27      Sunday  Browser   09:39:14           9   
7    51056  10-03-2019 2019-10-03    Thursday      App   01:24:23           1   

  event_timezone event_type  product_id  category_id     category  \
3            UTC   Purchase       10566           50  Electronics   
4            UTC       View        1924          115    Computers   
5            UTC       View         798           71  Electronics   
6            UTC       View         956           71  Electronics   
7            UTC       View       10616           50  Electronics   

  sub_category1 sub_category2    brand   price  user_session State  User_Score  
3         audio     headphone    Apple  169.35      11408172    ND           1  
4      notebook      notebook   Lenovo  851.59       3917052    MD           2  
5    smartphone    smartphone  Oneplus  566.04      12230929    MS           1  
6    smartphone    smartphone  Samsung  192.80      10947626    OH           3  
7         audio     headphone   Xiaomi   69.35      12349371    NC           2

In [50]:
'''
    Here i have calculated cr for each promoted product on its non-promotion days 
    followed by that i calculated the aggregated/avg cr for the same product and that is our baseline cr for that product
    which is then compared with promotion day cr to analyse the impact.
'''

def fun1(x):
    c=0
    for event in x:
        if event=='Purchase':
            c+=1
    return c/len(x)

base_cr_each_product={}    
for p_id in promo.ProductId:
    datee=promo[promo.ProductId==p_id].Date.iloc[0]
    base_cr_each_product[p_id]=baseline_cr[(baseline_cr.event_date != datee) & (baseline_cr.product_id==p_id)].groupby('event_date').agg(CR=('event_type',fun1))['CR'].mean()


In [51]:
baseline_cr_final=pd.DataFrame(base_cr_each_product.values(),index=base_cr_each_product.keys()).reset_index()
baseline_cr_final.columns=['product_id','Base_CR']
baseline_cr_final

product_id   Base_CR
0        10566  0.183748
1         1924  0.066667
2          798  0.047917
3          956  0.074573
4        10616  0.000000
..         ...       ...
56        3962  0.183761
57        6165  0.149008
58         871  0.153113
59        7431  0.125000
60         846  0.024479

[61 rows x 2 columns]

### Promotion day CR for each promoted product

In [52]:
# Conversion Rate of Promoted product on promotion day
def fun1(x):
    c=0
    for event in x:
        if event=='Purchase':
            c+=1
    return c/len(x)        

prod_level_CR=promoted_events.groupby(['event_date','product_id']).agg(CR=('event_type',fun1)).reset_index()
prod_level_CR=prod_level_CR.sort_values(by= 'event_date')
prod_level_CR

event_date  product_id        CR
0  2019-10-01     10566.0  0.434783
1  2019-10-06      6117.0  0.000000
2  2019-10-09       639.0  0.157895
3  2019-10-11     11618.0  0.000000
4  2019-10-12       564.0  0.333333
5  2019-10-13      5630.0  0.000000
6  2019-10-14     10605.0  0.100000
7  2019-10-15       896.0  0.333333
8  2019-10-16       885.0  0.000000
9  2019-10-18       915.0  0.000000
10 2019-10-19       448.0  0.160000
11 2019-10-20       691.0  0.400000
12 2019-10-26      1565.0  0.000000
13 2019-10-28      2124.0  0.000000
14 2019-10-29      3946.0  0.500000
15 2019-10-30       717.0  0.333333
16 2019-10-31       300.0  0.000000
17 2019-11-02       635.0  0.000000
18 2019-11-05      7552.0  0.000000
19 2019-11-10       874.0  0.166667
20 2019-11-11      7705.0  0.400000
21 2019-11-12      7108.0  0.000000
22 2019-11-13        11.0  0.000000
23 2019-11-14       976.0  0.000000
24 2019-11-15       577.0  0.000000
25 2019-11-16     11017.0  0.200000
26 2019-11-19      8612.0  0.000000
27 2019-11-21      1678.0  0.500000
28 2019-11-22       652.0  0.000000
29 2019-11-23       817.0  0.000000
30 2019-11-24      6155.0  0.000000
31 2019-11-25      7149.0  0.090909
32 2019-11-26      3962.0  0.500000
33 2019-11-27      6165.0  0.000000
34 2019-11-28       871.0  0.169811
35 2019-11-30       846.0  0.000000

In [112]:
# calculating revenue from promoted products
def fun(grp):
    grp=grp[grp.event_type=='Purchase']
    return grp.price.sum()
prod_level_rev=promoted_events.groupby(['event_date','product_id']).apply(fun).reset_index()
prod_level_cr_rev=prod_level_CR.merge(prod_level_rev,on=['event_date','product_id'],how='inner')
prod_level_cr_rev.rename(columns={0:'Total Revenue'},inplace=True)

C:\Users\welcome\AppData\Local\Temp\ipykernel_16032\719906992.py:5: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  prod_level_rev=promoted_events.groupby(['event_date','product_id']).apply(fun).reset_index()


In [113]:
prod_level_cr_rev

event_date  product_id        CR  Total Revenue
0  2019-10-01     10566.0  0.434783        1619.30
1  2019-10-06      6117.0  0.000000           0.00
2  2019-10-09       639.0  0.157895        1505.70
3  2019-10-11     11618.0  0.000000           0.00
4  2019-10-12       564.0  0.333333         169.32
5  2019-10-13      5630.0  0.000000           0.00
6  2019-10-14     10605.0  0.100000          23.02
7  2019-10-15       896.0  0.333333        1541.43
8  2019-10-16       885.0  0.000000           0.00
9  2019-10-18       915.0  0.000000           0.00
10 2019-10-19       448.0  0.160000        2984.88
11 2019-10-20       691.0  0.400000        1347.85
12 2019-10-26      1565.0  0.000000           0.00
13 2019-10-28      2124.0  0.000000           0.00
14 2019-10-29      3946.0  0.500000         396.48
15 2019-10-30       717.0  0.333333        1937.04
16 2019-10-31       300.0  0.000000           0.00
17 2019-11-02       635.0  0.000000           0.00
18 2019-11-05      7552.0  0.000000           0.00
19 2019-11-10       874.0  0.166667         969.75
20 2019-11-11      7705.0  0.400000         137.98
21 2019-11-12      7108.0  0.000000           0.00
22 2019-11-13        11.0  0.000000           0.00
23 2019-11-14       976.0  0.000000           0.00
24 2019-11-15       577.0  0.000000           0.00
25 2019-11-16     11017.0  0.200000         794.62
26 2019-11-19      8612.0  0.000000           0.00
27 2019-11-21      1678.0  0.500000         669.23
28 2019-11-22       652.0  0.000000           0.00
29 2019-11-23       817.0  0.000000           0.00
30 2019-11-24      6155.0  0.000000           0.00
31 2019-11-25      7149.0  0.090909         308.86
32 2019-11-26      3962.0  0.500000         450.46
33 2019-11-27      6165.0  0.000000           0.00
34 2019-11-28       871.0  0.169811        8247.33
35 2019-11-30       846.0  0.000000           0.00

### Comparison

In [114]:
impact_analysis_final=baseline_cr_final.merge(prod_level_cr_rev, how='left', on='product_id')
impact_analysis_final


product_id   Base_CR event_date        CR  Total Revenue
0        10566  0.183748 2019-10-01  0.434783        1619.30
1         1924  0.066667        NaT       NaN            NaN
2          798  0.047917        NaT       NaN            NaN
3          956  0.074573        NaT       NaN            NaN
4        10616  0.000000        NaT       NaN            NaN
..         ...       ...        ...       ...            ...
56        3962  0.183761 2019-11-26  0.500000         450.46
57        6165  0.149008 2019-11-27  0.000000           0.00
58         871  0.153113 2019-11-28  0.169811        8247.33
59        7431  0.125000        NaT       NaN            NaN
60         846  0.024479 2019-11-30  0.000000           0.00

[61 rows x 5 columns]

In [115]:
def fun(row):
    return row.unique()[0]

temp=promoted_events.groupby('product_id').agg(Brand=('brand',fun),Category=('category',fun)).reset_index()
impact_analysis_final=temp.merge(impact_analysis_final,how='right',on='product_id')
impact_analysis_final

product_id    Brand     Category   Base_CR event_date        CR  \
0      10566.0    Apple  Electronics  0.183748 2019-10-01  0.434783   
1       1924.0      NaN          NaN  0.066667        NaT       NaN   
2        798.0      NaN          NaN  0.047917        NaT       NaN   
3        956.0      NaN          NaN  0.074573        NaT       NaN   
4      10616.0      NaN          NaN  0.000000        NaT       NaN   
..         ...      ...          ...       ...        ...       ...   
56      3962.0       Lg  Electronics  0.183761 2019-11-26  0.500000   
57      6165.0       Lg   Appliances  0.149008 2019-11-27  0.000000   
58       871.0    Apple  Electronics  0.153113 2019-11-28  0.169811   
59      7431.0      NaN          NaN  0.125000        NaT       NaN   
60       846.0  Samsung  Electronics  0.024479 2019-11-30  0.000000   

    Total Revenue  
0         1619.30  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
..            ...  
56         450.46  
57           0.00  
58        8247.33  
59            NaN  
60           0.00  

[61 rows x 7 columns]

- The nan values of CR are corresponding to those 25 products which couldnt gather customer attention as also mentioned above.


In [116]:
with pd.ExcelWriter('Data_summary.xlsx') as writer:
    impact_analysis_final.to_excel(writer, sheet_name='Sheet1', index=False)
    Event_and_TotalPromo_summary.to_excel(writer, sheet_name='Sheet2', index=False)
    brand_prodcat_wise_interactions.to_excel(writer,sheet_name='sheet3',index=False)

In [82]:
cr_final= impact_analysis_final.dropna()
cr_final= cr_final[cr_final['CR']>cr_final['Base_CR']]
cr_final

product_id    Brand     Category   Base_CR event_date        CR
0      10566.0    Apple  Electronics  0.183748 2019-10-01  0.434783
11       564.0   Huawei  Electronics  0.121408 2019-10-12  0.333333
14       896.0    Apple  Electronics  0.231250 2019-10-15  0.333333
18       448.0    Apple  Electronics  0.150019 2019-10-19  0.160000
19       691.0  Samsung  Electronics  0.161354 2019-10-20  0.400000
28      3946.0       Lg  Electronics  0.053618 2019-10-29  0.500000
29       717.0  Samsung  Electronics  0.165531 2019-10-30  0.333333
40       874.0    Apple  Electronics  0.141193 2019-11-10  0.166667
41      7705.0  Samsung   Appliances  0.173183 2019-11-11  0.400000
46     11017.0    Apple  Electronics  0.123498 2019-11-16  0.200000
51      1678.0     Acer    Computers  0.081395 2019-11-21  0.500000
56      3962.0       Lg  Electronics  0.183761 2019-11-26  0.500000
58       871.0    Apple  Electronics  0.153113 2019-11-28  0.169811

Final Observations based on CR:

-   Out of 61 products promoted 25 did not show any type of traffic i.e no view,cart and purchase event occured.
-   13 (shown above) have shown significant impact.
-   The rest (61-25-13) 23 products had no significant impact as their CR is lower than the base CR.


In [98]:
# Promotion IDs of the promotion that showed significant impact
impact_analysis_final[impact_analysis_final['CR']>impact_analysis_final['Base_CR']].merge(promo, how='inner',left_on=['product_id','event_date'],\
        right_on=['ProductId','Date'])['Promotion Id'].unique()



array(['NzjwLJ2016', 'c1JgDc2016', 'APnD_x2016', 'EN4jLJ2016',
       'GWQDv82016', 'eJML432016', 'e3D1O-2016', 'lgsOJE2016'],
      dtype=object)

- The promotion strategy used corrsponding to these promotion ids should be applied again in future considering their positive impact on conversion rate.